src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,22 +1865,102 @@ public sealed class One : Tests
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    whenTrue = trueDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : trueDecisionPermitsTrueOther ? this : makeAndSequence(Not.Create(lengthTest), this);
+                    whenFalse = falseDecisionImpliesTrueOther ? makeOrSequence(lengthTest, this) : falseDecisionPermitsTrueOther ? this : makeAndSequence(Not.Create(lengthTest), this);
+
+                    static Tests makeOrSequence(Tests t1, Tests.One t2)
+                    {
+                        if (t1 is True) return t1;
+                        var builder = ArrayBuilder<Tests>.GetInstance(2);
+                        builder.Add(t1);
+                        builder.Add(t2);
+                        return OrSequence.Create(builder);
+                    }
+                    static Tests makeAndSequence(Tests t1, Tests.One t2)
+                    {
+                        if (t1 is False) return t1;
+                        var builder = ArrayBuilder<Tests>.GetInstance(2);
+                        builder.Add(t1);
+                        builder.Add(t2);
+                        return AndSequence.Create(builder);
+                    }
                 }
                 public override BoundDagTest ComputeSelectedTest() => this.Test;
                 public override Tests RemoveEvaluation(BoundDagEvaluation e) => e.Equals(Test) ? Tests.True.Instance : (Tests)this;
                 public override string Dump(Func<BoundDagTest, string> dump) => dump(this.Test);
                 public override bool Equals(object? obj) => this == obj || obj is One other && this.Test.Equals(other.Test);
                 public override int GetHashCode() => this.Test.GetHashCode();
+                public override Tests RewriteNestedLengthTests()
+                {
+                    BoundDagTest test = Test;
+                    if (test.Input.Source is BoundDagPropertyEvaluation { IsLengthOrCount: true } e)
+                    {
+                        if (TryGetTopLevelLengthTemp(e) is (BoundDagTemp lengthTemp, int offset))
+                        {
+                            switch (test)
+                            {
+                                case BoundDagValueTest t when !t.Value.IsBad:
```

Comment:
It would be good to have tests showing evaluation order. In some cases, we'll just call the top-level `Length`, but in other cases we'll also call the `Length` on the slice result.

Reply:
> but in other cases we'll also call the Length on the slice result.

No, it's always gets rewritten.

Reply:
Then let's show that (no Length call on the slice result).

Reply:
See https://github.com/dotnet/roslyn/pull/53891/files#r688391499